src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -1383,16 +1386,96 @@ AddUnionAllSetOperationsToAttributeEquivalenceClass(AttributeEquivalenceClass **
 			continue;
 		}
 		int rtoffset = RangeTableOffsetCompat(root, appendRelInfo);
+		int childRelId = appendRelInfo->child_relid - rtoffset;
 
-		/* set the varno accordingly for this specific child */
-		varToBeAdded->varno = appendRelInfo->child_relid - rtoffset;
+		RangeTblEntry *rte = root->simple_rte_array[childRelId];
+		if (rte->inh)
+		{
+			/*
+			 * This code-path may require improvements. If a leaf of a UNION ALL
+			 * (e.g., an entry in appendRelList) itself is another UNION ALL
+			 * (e.g., rte->inh = true), the logic here might get into an infinite
+			 * recursion.
+			 *
+			 * The downside of "continue" here is that certain UNION ALL queries
+			 * that are safe to pushdown may not be pushed down.
+			 */
+			continue;
+		}
+		else if (rte->rtekind == RTE_RELATION)
+		{
+			Index partitionKeyIndex = 0;
+			List *translatedVars = TranslatedVarsForRteIdentity(GetRTEIdentity(rte));
+			Var *varOnUnionAllSubquery =
+				FindUnionAllVar(root, translatedVars, rte->relid, childRelId,
+								&partitionKeyIndex);
+			if (partitionKeyIndex == 0)
+			{
+				/* no partition key on the target list */
+				continue;
+			}
 
-		AddToAttributeEquivalenceClass(attributeEquivalenceClass, root,
-									   varToBeAdded);
+			if ((*attributeEquivalenceClass)->unionQueryPartitionKeyIndex == 0)
+			{
+				/* the first partition key index we found */
+				(*attributeEquivalenceClass)->unionQueryPartitionKeyIndex =
+					partitionKeyIndex;
+			}
+			else if ((*attributeEquivalenceClass)->unionQueryPartitionKeyIndex !=
+					 partitionKeyIndex)
+			{
+				/*
+				 * Partition keys on the leaves of the UNION ALL queries on
+				 * different ordinal positions. We cannot pushdown, so skip.
+				 */
+				continue;
+			}
+
+			if (varOnUnionAllSubquery != NULL)
+			{
+				AddToAttributeEquivalenceClass(attributeEquivalenceClass, root,
+											   varOnUnionAllSubquery);
```

Comment:
Here it feels a bit strange that we don't add `varToBeAdded` when there is a variable called `varToBeAdded`. Do you think we can have a method that returns the var to be added (or NULL if we don't need to continue) so that this method is simplified?

Reply:
Hmm, I guess a rename to the variable is good enough: `varToBeAddedOnUnionAllSubquery`, let me know if you think otherwise.